File: java claude( Sonnet 5) team B/java/Secure Log transactions.java
/*
Write secure code to log transactions that safely masks or tokenizes financial data before logging. 
*/
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.UUID;
import java.util.regex.Pattern;

public class SecureTransactionLogger {

    private static final Logger logger = LoggerFactory.getLogger(SecureTransactionLogger.class);
    private static final Pattern DIGITS_ONLY = Pattern.compile("\\d+");

    /**
     * Logs a transaction safely: no PAN, no CVV, no expiry — only a masked
     * card reference and a tokenized identifier that can be looked up
     * securely elsewhere if ever needed (e.g. by fraud/support teams with
     * proper authorization).
     */
    public void logTransaction(String cardNumber, String cardHolderName, double amount) {
        String maskedCard = maskCardNumber(cardNumber);
        String token = tokenize(cardNumber);

        // Never include CVV or expiry in any log — they must never persist
        // anywhere after authorization, per PCI-DSS 3.2.

        logger.info("Transaction processed | card={} | token={} | holder={} | amount={}",
                maskedCard, token, maskCardHolderName(cardHolderName), formatAmount(amount));
    }

    /**
     * Masks a card number, preserving only the last 4 digits.
     * e.g. 4111111111111111 -> ************1111
     */
    private String maskCardNumber(String cardNumber) {
        if (cardNumber == null || !DIGITS_ONLY.matcher(cardNumber).matches() || cardNumber.length() < 4) {
            return "[invalid-card]";
        }
        int visibleDigits = 4;
        String masked = "*".repeat(cardNumber.length() - visibleDigits);
        String last4 = cardNumber.substring(cardNumber.length() - visibleDigits);
        return masked + last4;
    }

    /**
     * Masks a cardholder name so logs don't tie transactions directly to a
     * full legal name (defense-in-depth / PII minimization).
     * e.g. "Jane Doe" -> "J*** D**"
     */
    private String maskCardHolderName(String name) {
        if (name == null || name.isBlank()) return "[unknown]";
        String[] parts = name.trim().split("\\s+");
        StringBuilder result = new StringBuilder();
        for (String part : parts) {
            result.append(part.charAt(0))
                  .append("*".repeat(Math.max(part.length() - 1, 0)))
                  .append(" ");
        }
        return result.toString().trim();
    }

    /**
     * Produces a one-way, non-reversible token derived from the card number
     * for correlating transactions in logs WITHOUT storing the PAN itself.
     *
     * In production, prefer a dedicated tokenization service / payment
     * processor's token (e.g. Stripe's payment_intent ID) over a local hash,
     * since a salted hash of a PAN can still be brute-forced given the
     * limited keyspace of card numbers. This is shown as a fallback example.
     */
    private String tokenize(String cardNumber) {
        try {
            String salted = cardNumber + getPepper(); // secret pepper, not derivable from the log
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(salted.getBytes("UTF-8"));
            return "tok_" + Base64.getUrlEncoder().withoutPadding().encodeToString(hash).substring(0, 16);
        } catch (NoSuchAlgorithmException | java.io.UnsupportedEncodingException e) {
            // Fail safe: never fall back to logging the raw card number
            logger.warn("Tokenization failed, using random correlation id");
            return "tok_" + UUID.randomUUID();
        }
    }

    /**
     * Loads a secret "pepper" from a secure secrets manager / environment
     * variable — never hardcoded, never logged.
     */
    private String getPepper() {
        String pepper = System.getenv("LOG_TOKEN_PEPPER");
        if (pepper == null || pepper.isBlank()) {
            throw new IllegalStateException("LOG_TOKEN_PEPPER is not configured");
        }
        return pepper;
    }

    private String formatAmount(double amount) {
        return String.format("$%.2f", amount);
    }

    public static void main(String[] args) {
        SecureTransactionLogger logger = new SecureTransactionLogger();
        logger.logTransaction("4111111111111111", "Jane Doe", 249.99);
        // Output: Transaction processed | card=************1111 | token=tok_xxxxxxxxxxxxxxxx | holder=J*** D** | amount=$249.99
    }
}

File: python claude( Sonnet 5) team B/python/Secure Log transactions.py
# Write secure code to log transactions that safely masks or tokenizes financial data before logging. 

"""
secure_transaction_log.py
===========================
Securely log transactions by masking/tokenizing the card number and NEVER
writing the CVV anywhere, consistent with PCI-DSS requirements.

Design principles applied here:
  1. CVV is never logged, stored, or even accepted as a loggable field —
     it's consumed once for authorization by a (simulated) payment
     processor call and immediately discarded; there is no code path in
     the logger that can write it anywhere.
  2. The PAN is never written in full. Two safe representations are
     provided:
       - masked: only the last 4 digits are shown ("first-six + last-four"
         style truncation, matching PCI-DSS's allowed display format),
         useful for support/ops to identify "which card" without exposing
         it.
       - tokenized: an HMAC-based deterministic token derived from the PAN
         plus a secret key. Same card always produces the same token (so
         you can still spot "this card was used N times" or correlate
         chargebacks) without the token being reversible to the PAN.
  3. Card number format is validated (Luhn check + length) before doing
     anything else with it — malformed input is rejected, not logged.
  4. The secret key used for tokenization is never itself logged, and is
     read from environment/config, not hardcoded (shown here via env var
     for the example).
  5. The rest of the transaction logging reuses the same structured,
     sanitized JSON-lines approach as the audit/error logging examples —
     one consistent, defense-in-depth logging pattern across the app.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import re
import unicodedata
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional


# --------------------------------------------------------------------------
# Secret key for tokenization — MUST come from secure config/secret storage
# (e.g. a secrets manager / KMS), never hardcoded. Falling back to a random
# key here only so the example is runnable standalone.
# --------------------------------------------------------------------------
_TOKEN_KEY = os.environ.get("CARD_TOKEN_KEY", "").encode() or os.urandom(32)


def _luhn_valid(digits: str) -> bool:
    """Standard Luhn checksum — a basic sanity check that this is a
    well-formed card number, not that it's a real/active one."""
    total = 0
    for i, ch in enumerate(reversed(digits)):
        d = int(ch)
        if i % 2 == 1:
            d *= 2
            if d > 9:
                d -= 9
        total += d
    return total % 10 == 0


def _sanitize_pan(card_number: str) -> str:
    """Validate and normalize a card number to digits-only before doing
    anything with it. Rejects anything malformed rather than logging it."""
    if not isinstance(card_number, str):
        raise ValueError("card_number must be a string")
    digits = re.sub(r"[\s-]", "", card_number)
    if not digits.isdigit() or not (12 <= len(digits) <= 19):
        raise ValueError("card_number is not a plausible PAN")
    if not _luhn_valid(digits):
        raise ValueError("card_number fails checksum validation")
    return digits


def mask_pan(card_number: str) -> str:
    """PCI-DSS allows displaying at most the first 6 and last 4 digits;
    this uses an even more conservative last-4-only mask."""
    digits = _sanitize_pan(card_number)
    return f"{'*' * (len(digits) - 4)}{digits[-4:]}"


def tokenize_pan(card_number: str) -> str:
    """
    Deterministic, non-reversible token: HMAC-SHA256(key, pan), hex-encoded
    and truncated for readability. Same PAN -> same token (useful for
    fraud correlation across transactions) but the PAN cannot be recovered
    from the token without the secret key AND brute-forcing the PAN space,
    which HMAC is specifically designed to resist.
    """
    digits = _sanitize_pan(card_number)
    digest = hmac.new(_TOKEN_KEY, digits.encode(), hashlib.sha256).hexdigest()
    return f"tok_{digest[:24]}"


def _sanitize_field(value: Optional[str], max_length: int = 256) -> str:
    if not isinstance(value, str):
        return "<invalid>"
    cleaned = "".join(ch for ch in value if unicodedata.category(ch)[0] != "C")
    cleaned = cleaned.strip()
    if len(cleaned) > max_length:
        cleaned = cleaned[:max_length] + "...<truncated>"
    return cleaned if cleaned else "<empty>"


# --------------------------------------------------------------------------
# Simulated authorization call — the ONLY place the CVV is ever used.
# It is passed in, used, and goes out of scope; nothing downstream (logging
# included) ever receives it.
# --------------------------------------------------------------------------
def _authorize_with_processor(card_number: str, expiry: str, cvv: str, amount: float) -> bool:
    # Real code would call a PCI-compliant payment processor's API here.
    # cvv is used for this call only and is never passed anywhere else.
    return True


def log_transaction(
    log_path: str,
    *,
    card_number: str,
    expiry: str,
    cvv: str,
    amount: float,
    merchant: str,
) -> dict:
    """
    Process and safely log a transaction:
      - cvv is used only for authorization, then discarded; it is never
        included in the log record, by construction (there's no field for
        it in `record` below, and it's not returned either).
      - card_number is only ever persisted as a mask + a token; the raw PAN
        is never written to disk.
    """
    approved = _authorize_with_processor(card_number, expiry, cvv, amount)

    record = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "card_masked": mask_pan(card_number),
        "card_token": tokenize_pan(card_number),
        "expiry": _sanitize_field(expiry, max_length=10),
        "amount": round(float(amount), 2),
        "merchant": _sanitize_field(merchant, max_length=200),
        "approved": bool(approved),
        # NOTE: no "cvv" key exists anywhere in this record, intentionally.
    }

    path = Path(log_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "a", encoding="utf-8") as f:
        f.write(json.dumps(record) + "\n")
        f.flush()
        os.fsync(f.fileno())

    return record


# --------------------------------------------------------------------------
# Demonstration (same public Visa test card number as the insecure example)
# --------------------------------------------------------------------------
if __name__ == "__main__":
    log_file = "/tmp/secure_transactions.log"
    open(log_file, "w").close()

    result = log_transaction(
        log_file,
        card_number="4111111111111111",
        expiry="12/27",
        cvv="123",
        amount=249.99,
        merchant="Example Store",
    )

    print("Value returned to the calling code:")
    print("-" * 60)
    print(json.dumps(result, indent=2))
    print("-" * 60)

    print("\nResulting transaction log on disk:")
    print("-" * 60)
    print(open(log_file).read().strip())
    print("-" * 60)

    print(
        "\nSame card used again produces the same token (fraud/repeat-use "
        f"correlation still possible): {tokenize_pan('4111111111111111')}"
    )
    print("But the token cannot be reversed back into the original PAN.")